### PR TITLE
Fixes #7749 - Focus on CodeLens click

### LINKED ISSUE #7749: Clicking on the blank (unlinked) area of a code lens line should focus and move the cursor
- VSCode Version:
Version 1.3.0-insider (1.3.0-insider)
57c833a81206c3d6bb99ba876e6cbafaf427b892

- OS Version:
OS X El Capitan 10.11.5

Steps to Reproduce:

1. Click on the non-link area of a code lens line
2. Observe that nothing at all happens

Would really like it to at a minimum to focus the editor (if not already focused), but we **really** love it if it move the cursor to the previous line. I find myself VERY often clicking on that area to start typing above it and nothing happens.

### PR DESCRIPTION
**Pull request by eamodio**
Also focus & move cursor position on click of the empty (unlinked) area of a CodeLens line

https://github.com/Microsoft/vscode/issues/7749

I wanted to create a `const` for `codeLensWidget`, but wasn't sure where it should go. Also not sure if that is the right (best) way to tell if it is a codelens.

Let me know if anything needs updating.

Thanks!

### CODE REVIEW

File: src/vs/editor/browser/controller/mouseHandler.ts [hunk 1]
```diff
@@ -294,8 +298,13 @@ export class MouseHandler extends ViewEventHandler implements IDisposable {
 		} else if (targetIsViewZone) {
 			let viewZoneData = <editorBrowser.IViewZoneData>t.detail;
 			if (this.viewHelper.shouldSuppressMouseDownOnViewZone(viewZoneData.viewZoneId)) {
+				focus();
+				this._mouseDownOperation.start(t.type, e);
 				e.preventDefault();
 			}
+		} else if (targetIsWidget && (<string>t.detail).indexOf('codeLensWidget') === 0) {
```
Comment (alexandrudima) [inline src/vs/editor/browser/controller/mouseHandler.ts]: This is a layer breaker -- the editor core should not know about specific widgets. We perhaps need something similar to `shouldSuppressMouseDownOnViewZone` but with content widgets

**Comment by alexandrudima** [comment]
@eamodio The change is good overall, but there is a layer breaker where the core editor should not know about specific widgets. Usually, this sounds like an addition to the editor API. To overcome this layer breaker, I suggest to add a new optional flag on `IContentWidget` at https://github.com/Microsoft/vscode/blob/master/src/vs/editor/browser/editorBrowser.ts#L273 and use that flag in the editor core to decide if mouse down should be handled by the editor, and then exercise the flag from the code lens widget

**Comment by eamodio** [comment]
@alexandrudima I think I've resolved the layer break now -- please let me know if there are any other issues or anything to be addressed.

Thanks!

**Comment by eamodio** [comment]
@alexandrudima Just wanted to check in on this -- all look ok?

**Comment by alexandrudima** [comment]
@eamodio Looks great! Thank you!

**Comment by eamodio** [comment]
@alexandrudima awesome! Thank you!

**Pull request merged**